lib/galaxy/webapps/galaxy/services/visualizations.py

```diff
@@ -60,3 +97,360 @@ def index(
+        visualization_id: DecodedDatabaseIdField,
+    ) -> VisualizationShowResponse:
+        """Return a dictionary containing the Visualization's details
+
+        :rtype:     dictionary
+        :returns:   Visualization details
+        """
+        # TODO: revisions should be a contents/nested controller like viz/xxx/r/xxx)?
+        # the important thing is the config
+        # TODO:?? /api/visualizations/registry -> json of registry.listings?
+        visualization = self._get_visualization(trans, visualization_id, check_ownership=False, check_accessible=True)
+        dictionary = {
+            "model_class": "Visualization",
+            "id": visualization.id,
+            "title": visualization.title,
+            "type": visualization.type,
+            "user_id": visualization.user.id,
+            "dbkey": visualization.dbkey,
+            "slug": visualization.slug,
+            # to_dict only the latest revision (allow older to be fetched elsewhere)
+            "latest_revision": (
+                self._get_visualization_revision_dict(visualization.latest_revision)
+                if visualization.latest_revision
+                else None
+            ),
+            # need to encode ids in revisions as well
+            # NOTE: does not encode ids inside the configs
+            "revisions": [r.id for r in visualization.revisions],
+        }
+        # replace with trans.url_builder if possible
+        dictionary["url"] = url_for(
+            controller="visualization",
+            action="display_by_username_and_slug",
+            username=visualization.user.username,
+            slug=visualization.slug,
+        )
+        dictionary["username"] = visualization.user.username
+        dictionary["email_hash"] = md5_hash_str(visualization.user.email)
+        dictionary["tags"] = visualization.make_tag_string_list()
+        dictionary["annotation"] = get_item_annotation_str(trans.sa_session, trans.user, visualization)
+        app = cast(StructuredApp, trans.app)
+        if app.visualizations_registry:
+            visualizations_registry = cast(VisualizationsRegistry, app.visualizations_registry)
+            visualization_plugin = cast(VisualizationPlugin, visualizations_registry.get_plugin(dictionary["type"]))
+            dictionary["plugin"] = visualization_plugin.to_dict()
+        return VisualizationShowResponse(**dictionary)
+
+    def create(
+        self,
+        trans: ProvidesUserContext,
+        payload: VisualizationCreatePayload,
+    ) -> VisualizationCreateResponse:
+        """Returns a dictionary of the created visualization
+
+        :rtype:     dictionary
+        :returns:   dictionary containing Visualization details
+        """
+
+        if payload.import_id:
+            visualization = self._import_visualization(trans, payload.import_id)
+        else:
+            payload = cast(VisualizationCreatePayload, self._validate_and_parse_payload(payload))
```

Comment:
This doesn't look correct, `_validate_and_parse_payload` is what the Pydantic model is doing right now, so no need for revalidation if the schema model is correctly typed. You might need to define a custom validator to sanitize the HTML, and assign that type to those fields that require it. Have a look at https://docs.pydantic.dev/latest/concepts/validators/#annotated-validators

Reply:
Is it ok to use something like this in `VisualizationUpdatePayload` and `VisualizationCreatePayload`:
```python
    @field_validator("type", "title", "dbkey", "slug", "annotation", mode="before")
    @classmethod
    def sanitize_html_fields(cls, v):
        if isinstance(v, str):
            return sanitize_html(v)
        return v
```
I have seen them in `SearchJobsPayload`, `BaseDataPayload`, and `InvokeWorkflowPayload`.

Reply:
I think that should work, yes, although I would slightly prefer something more reusable (at field type level) that you don't need to duplicate from model to model.

Something like a type with custom validation called `SanitizedString` that you can assign to any field that requires it or something similar. But that is up to you, both are valid as long as the fields get sanitized.

Reply:
Something like this?
```python
class SanitizedString(str):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, value):
        if isinstance(value, str):
            return cls(sanitize_html(value))
        raise TypeError("string required")

    @classmethod
    def __get_pydantic_core_schema__(cls, source_type, handler):
        return core_schema.no_info_after_validator_function(
            cls.validate,
            core_schema.str_schema(),
            serialization=core_schema.to_string_ser_schema(),
        )
```